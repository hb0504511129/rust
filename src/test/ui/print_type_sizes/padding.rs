// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// compile-flags: -Z print-type-sizes

// This file illustrates how padding is handled: alignment
// requirements can lead to the introduction of padding, either before
// fields or at the end of the structure as a whole.

#![allow(dead_code)]

struct S {
    a: bool,
    b: bool,
    g: i32,
}

enum E1 {
    A(i64, i32),
    B(S),
}

enum E2 {
    A(i32, i64),
    B(S),
}

fn main() { }
