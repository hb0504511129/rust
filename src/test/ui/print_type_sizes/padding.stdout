print-type-size type: `E1`: 24 bytes, alignment: 8 bytes
print-type-size     discriminant: 4 bytes
print-type-size     variant `A`: 16 bytes
print-type-size         padding bytes: 4
print-type-size         field `.0`: 8 bytes, alignment: 8 bytes
print-type-size         field `.1`: 4 bytes
print-type-size     variant `B`: 8 bytes
print-type-size         field `.0`: 8 bytes
print-type-size     end padding bytes: 4
print-type-size type: `E2`: 16 bytes, alignment: 8 bytes
print-type-size     discriminant: 4 bytes
print-type-size     variant `A`: 12 bytes
print-type-size         field `.0`: 4 bytes
print-type-size         field `.1`: 8 bytes
print-type-size     variant `B`: 8 bytes
print-type-size         field `.0`: 8 bytes
print-type-size type: `S`: 8 bytes, alignment: 4 bytes
print-type-size     field `.a`: 1 bytes
print-type-size     field `.b`: 1 bytes
print-type-size     padding bytes: 2
print-type-size     field `.g`: 4 bytes, alignment: 4 bytes
